File: stone_processor/src/main/java/com/github/klee0kai/stone/helpers/itemholder/ItemCacheType.java
package com.github.klee0kai.stone.helpers.itemholder;

import com.github.klee0kai.stone._hidden_.types.holders.StRefType;
import com.github.klee0kai.stone.annotations.component.GcSoftScope;
import com.github.klee0kai.stone.annotations.component.GcStrongScope;
import com.github.klee0kai.stone.annotations.component.GcWeakScope;
import com.github.klee0kai.stone.annotations.module.BindInstance;
import com.github.klee0kai.stone.annotations.module.Provide;
import com.squareup.javapoet.ClassName;

public enum ItemCacheType {
    Strong, Soft, Weak;

    public ClassName getGcScopeClassName() {
        switch (this) {
            case Weak:
                return ClassName.get(GcWeakScope.class);
            case Strong:
                return ClassName.get(GcStrongScope.class);
            case Soft:
            default:
                return ClassName.get(GcSoftScope.class);
        }
    }

    public static ItemCacheType cacheTypeFrom(BindInstance.CacheType cacheType) {
        if (cacheType != null) switch (cacheType) {
            case Weak:
                return ItemCacheType.Weak;
            case Strong:
                return ItemCacheType.Strong;
            case Soft:
            default:
                return ItemCacheType.Soft;
        }
        return ItemCacheType.Soft;
    }

    public static ItemCacheType cacheTypeFrom(Provide.CacheType cacheType) {
        if (cacheType != null) switch (cacheType) {
            case Weak:
                return ItemCacheType.Weak;
            case Strong:
                return ItemCacheType.Strong;
            case Soft:
            default:
                return ItemCacheType.Soft;
        }
        return ItemCacheType.Soft;
    }

    public StRefType toRefTypeSingle() {
        switch (this) {
            case Strong:
                return StRefType.StrongObject;
            case Soft:
                return StRefType.SoftObject;
            case Weak:
                return StRefType.WeakObject;
            default:
                return null;
        }
    }

    public StRefType toRefTypeList() {
        switch (this) {
            case Strong:
                return StRefType.ListObject;
            case Soft:
                return StRefType.ListSoftObject;
            case Weak:
                return StRefType.ListWeakObject;
            default:
                return null;
        }
    }

}

File: stone_ksp/src/commonMain/kotlin/com/github/klee0kai/thekey/stone/ksp/helpers/itemholder/ItemCacheType.kt
package com.github.klee0kai.thekey.stone.ksp.helpers.itemholder

import com.github.klee0kai.stone.__hidden__.types.holders.StoneRefType
import com.github.klee0kai.stone.annotations.component.GcSoftScope
import com.github.klee0kai.stone.annotations.component.GcStrongScope
import com.github.klee0kai.stone.annotations.component.GcWeakScope
import com.github.klee0kai.stone.annotations.module.BindInstance
import com.github.klee0kai.stone.annotations.module.Provide
import com.squareup.kotlinpoet.asClassName

enum class ItemCacheType {
    Strong, Soft, Weak;

    val gcScopeClassName
        get() = when (this) {
            Weak -> GcWeakScope::class
            Strong -> GcStrongScope::class
            Soft -> GcSoftScope::class
        }.asClassName()
}

fun ItemCacheType.toRefTypeSingle(

): StoneRefType = when (this) {
    ItemCacheType.Strong -> StoneRefType.StrongObject
    ItemCacheType.Soft -> StoneRefType.SoftObject
    ItemCacheType.Weak -> StoneRefType.WeakObject
}

fun ItemCacheType.toRefTypeList(

): StoneRefType = when (this) {
    ItemCacheType.Strong -> StoneRefType.ListObject
    ItemCacheType.Soft -> StoneRefType.ListSoftObject
    ItemCacheType.Weak -> StoneRefType.ListWeakObject
}


fun BindInstance.CacheType.toItemCacheType(
): ItemCacheType = when (this) {
    BindInstance.CacheType.Weak -> ItemCacheType.Weak
    BindInstance.CacheType.Soft -> ItemCacheType.Soft
    BindInstance.CacheType.Strong -> ItemCacheType.Strong
}

fun Provide.CacheType.toItemCacheType(
): ItemCacheType? = when (this) {
    Provide.CacheType.Factory -> null
    Provide.CacheType.Weak -> ItemCacheType.Weak
    Provide.CacheType.Soft -> ItemCacheType.Soft
    Provide.CacheType.Strong -> ItemCacheType.Strong
}

